### PR TITLE
[clangd] Allow specifying what headers are always included via "" or <>

### PR DESCRIPTION
**Pull request by kleinesfilmroellchen**
Projects can now add config fragments like this to their .clangd:

```yaml
Style:
  QuotedHeaders: "src/.*"
  AngledHeaders: ["path/sdk/.*", "third-party/.*"]
```

to force headers inserted via the --header-insertion=iwyu mode matching at least one of the regexes to have <> (AngledHeaders) or "" (QuotedHeaders) around them, respectively. For other headers (and in conflicting cases where both styles have a matching regex), the current system header detection remains.

Ref https://github.com/clangd/clangd/issues/1247

Based on https://reviews.llvm.org/D145843

This solution does not affect other clang tools like clang-format.

CC @HighCommander4 @adkaster. I have no merge rights so someone else needs to commit this.

### CODE REVIEW

**Comment by kleinesfilmroellchen** [comment]
I investigated the test failures some more in the context of where the bug affects real uses and I'm unsure as to what's going on.
```yaml
Style:
  QuotedHeaders: ["blahblahblah"]
  AngledHeaders: ["AK/.*", "Userland/.*", "Kernel/.*", "Applications/.*", "Lib.*/.*"]
```
will match all regexes correctly.
```yaml
Style:
  AngledHeaders: ["AK/.*", "Userland/.*", "Kernel/.*", "Applications/.*", "Lib.*/.*"]
```
will make AngledHeaders completely disappear from the parsed and compiled config; apparently the dictionary key doesn't even exist anymore. Exactly this behavior makes a test fail on CI.
```yaml
Style:
  QuotedHeaders: ["blahblahblah"]
```
will keep QuotedHeaders in the parsed and compiled config and works as expected.
There is no apparent difference in how these two keys are handled, I have cross-checked my config parser and compiler code with existing code countless times. Really no clue what's happening here other than QuotedHeaders apparently needs to exist or AngledHeaders vanishes.
CC @HighCommander4

File: clang-tools-extra/clangd/ConfigCompile.cpp [hunk 1]
```diff
@@ -483,6 +483,72 @@ struct FragmentCompiler {
...
+               llvm::formatv("Invalid regular expression '{0}': {1}",
+                             *HeaderPattern, RegexError)
+                   .str(),
+               HeaderPattern.Range);
+          continue;
+        }
+        Filters->push_back(std::move(CompiledRegex));
+      }
+      if (Filters->empty())
+        return;
```
Comment (zyn0217) [inline clang-tools-extra/clangd/ConfigCompile.cpp:511]: I think this is why you're seeing the puzzling behavior: You've bailed out of parsing the `QuotedHeaders` block if don't see any `AngledHeaders` field. Perhaps extract these two blocks into separate functions?

**Comment by kleinesfilmroellchen** [comment]
Thanks for the finding @zyn0217, refactored the config compiler code a little. Tests should pass now and I have re-tested this on SerenityOS and other small project where it works as expected in practice.

File: clang-tools-extra/clangd/ConfigCompile.cpp [hunk 2]
```diff
@@ -483,6 +483,56 @@ struct FragmentCompiler {
...
+      -> std::optional<std::function<bool(llvm::StringRef)>> {
+    // TODO: Share this code with Diagnostics.Includes.IgnoreHeader
+#ifdef CLANGD_PATH_CASE_INSENSITIVE
+    static llvm::Regex::RegexFlags Flags = llvm::Regex::IgnoreCase;
+#else
+    static llvm::Regex::RegexFlags Flags = llvm::Regex::NoFlags;
+#endif
+    auto Filters = std::make_shared<std::vector<llvm::Regex>>();
+    for (auto &HeaderPattern : HeaderPatterns) {
+      elog("found angle pattern {0}", *HeaderPattern);
```
Comment (ADKaster) [inline clang-tools-extra/clangd/ConfigCompile.cpp:512]: elog seems a bit high for this debugging output.

Comment (kleinesfilmroellchen) [reply clang-tools-extra/clangd/ConfigCompile.cpp:512]: Oops, that's not supposed to be there anymore 😅

**Comment by HighCommander4** [comment]
What is the relationship between this patch, and clangd 17's ["missing include" warning](https://clangd.llvm.org/guides/include-cleaner#missing-include-warning)? Does the quick-fix for the "missing include" warning also respect these config options?

**Comment by tudor** [comment]
> What is the relationship between this patch, and clangd 17's ["missing include" warning](https://clangd.llvm.org/guides/include-cleaner#missing-include-warning)? Does the quick-fix for the "missing include" warning also respect these config options?

Same question from me -- this is the main reason I personally care about this PR :) We have directories in our project tree that we'd like to be included as `#include <dir/foo.h>` (as those would be installed system-wide in an open-source version) and others that we'd like to be included as `#include "dir/foo.h"` (internal headers that are in the same subtree as the C++ source) and I'd really like clangd in vscode to suggest fixing the missing include warnings appropriately. Thanks!

**Comment by kleinesfilmroellchen** [comment]
> > What is the relationship between this patch, and clangd 17's ["missing include" warning](https://clangd.llvm.org/guides/include-cleaner#missing-include-warning)? Does the quick-fix for the "missing include" warning also respect these config options?
> 
> Same question from me -- this is the main reason I personally care about this PR :) We have directories in our project tree that we'd like to be included as `#include <dir/foo.h>` (as those would be installed system-wide in an open-source version) and others that we'd like to be included as `#include "dir/foo.h"` (internal headers that are in the same subtree as the C++ source) and I'd really like clangd in vscode to suggest fixing the missing include warnings appropriately. Thanks!

As far as I can tell, clangd uses the correct header style for inserted includes regardles of how those includes are created (via auto-include on code completion or fix suggestions).

**Comment by ADKaster** [comment]
@HighCommander4 @sam-mccall do you have any capacity to review this change? It would be nice to get it into the 18 release if the window hasn't closed.

**Comment by HighCommander4** [comment]
> > > What is the relationship between this patch, and clangd 17's ["missing include" warning](https://clangd.llvm.org/guides/include-cleaner#missing-include-warning)? Does the quick-fix for the "missing include" warning also respect these config options?
>
> As far as I can tell, clangd uses the correct header style for inserted includes regardles of how those includes are created (via auto-include on code completion or fix suggestions).

I played around with this a bit, and discovered that there are actually **three** situations in which clangd inserts a header:

 1. During code completion ("header insertion")
 2. In a quick-fix for an **error** diagnostic on an unresolved name ("include fixer")
 3. In a quick-fix for a **warning** diagnostic (technically, a diagnostic with severity [`DiagnosticSeverity.Information`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnosticSeverity)) on a name that's resolved, but where the file declaring it is not _directly_ included ("include cleaner")

The patch seems to handle the first two, but not the third.

The third scenario can be tested by creating a workspace with the following files:

.clangd:
```yaml
CompileFlags:
  Add: -I.
Diagnostics:
  MissingIncludes: Strict
Style:
  QuotedHeaders: "quoted/.*"
  AngledHeaders: "angled/.*"
```

quoted/quoted.hpp:
```c++
#ifndef QUOTED_HPP_
#define QUOTED_HPP_
void quoted();
#endif
```

angled/angled.hpp:
```c++
#ifndef ANGLED_HPP_
#define ANGLED_HPP_
void angled();
#endif
```

indirect.hpp:
```c++
#include "quoted/quoted.hpp"
#include <angled/angled.hpp>
```

main.cpp:
```c++
#include "indirect.hpp"
int main() {
  quoted();
  angled();
}
```

Now, in main.cpp, if you hover over the `angled` token and accept the quick fix proposed there, the added include is spelled `"angled/angled.hpp"` rather than `<angled/angled.hpp>`.

**Comment by HighCommander4** [comment]
> @HighCommander4 @sam-mccall do you have any capacity to review this change?

I'm happy to try and help, but I haven't really been involved in the development of the include-cleaner component, and based on the findings in my previous comment, the patch would benefit from feedback from someone more familiar with that, to provide guidance about where in the code the config option should be checked to handle the include-cleaner scenario as well.

**Comment by kleinesfilmroellchen** [comment]
After some discussion, I would like to land the feature "partially" as-is without support for "include cleaner" header insertion. (Case 3 in @HighCommander4's elaboration above.) This is motivated by two factors:
- Having looked at a way of supporting all three insertion cases, it seems like some significant refactoring is required. The common code between the three cases is either upstream in Clang (`HeaderSearch`), which would probably be a *very* leaky and unclean abstraction. Or downstream in replacement generation (`HeaderIncludes`), where the relevant information to make the decision is no longer available. I'm honestly not excited about performing such a refactor, as I am not particularly familiar with the complex code base.
- The other cases seem to be more common in practice, and have already been very useful in my dogfooding with SerenityOS. Within our project this patch would already be appreciated as a great improvement over the status quo.
If maintainers agree, I'd like to have this merged before 18.

File: clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp [hunk 3]
```diff
@@ -539,6 +539,42 @@ TEST_F(ConfigCompileTests, Style) {
...
+    Frag.Style.QuotedHeaders.push_back(Located<std::string>(".*inc"));
+    EXPECT_TRUE(compileAndApply());
+    auto HeaderFilter = [this](llvm::StringRef Path) {
+      for (auto &Filter : Conf.Style.QuotedHeaders) {
+        if (Filter(Path))
+          return true;
+      }
+      return false;
+    };
+    EXPECT_TRUE(HeaderFilter("foo.h"));
```
Comment (HighCommander4) [inline clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp:557]: consider checking `"prefix/foo.h"` as well

**Review by HighCommander4 — COMMENTED** [review]
I took a pass through the implementation. Generally looks good.

Would be nice to have a test that exercises the include insertion process in a more end-to-end way. I think something in `CodeCompleteTests.cpp` would work, using `WithContextValue` to set config values ([example](https://searchfox.org/llvm/rev/365aa1574a1b4a3cdee6648227d095d00536ffde/clang-tools-extra/clangd/unittests/InlayHintTests.cpp#77-84,100)), and using the [`insertIncludeText`](https://searchfox.org/llvm/rev/365aa1574a1b4a3cdee6648227d095d00536ffde/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp#83) matcher to make assertions about the spelling of inserted includes.

**Review by sam-mccall — COMMENTED** [review]
Sorry about inactivity on my part (here and in general). Thanks for working on this, and thanks @HighCommander4 as always - will leave LGTM to you.

Landing this without include-cleaner support is OK with me, though a bit sad.

I'm not sure I see the difficulty though:
IncludeCleaner.cpp has `generateMissingIncludeDiagnostics()` which already takes the IgnoreHeaders filter, and decides whether to spell with angles or quotes based on the include-cleaner library's suggested spelling (`Angled`). Can we just override that based on the QuotedHeaders etc filters?

That wouldn't validate whether the include can actually be included with the flipped quoting style, but it almost always is, and ISTM the other paths also don't validate this.

It's sad that there's not a single codepath that would handle all these, but maybe the perfect is the enemy of the good here.

File: clang-tools-extra/clangd/ConfigFragment.h [hunk 4]
```diff
@@ -296,6 +296,19 @@ struct Fragment {
     // ::). All nested namespaces are affected as well.
     // Affects availability of the AddUsing tweak.
     std::vector<Located<std::string>> FullyQualifiedNamespaces;
+
+    /// List of regexes for headers that should always be included with a
+    /// ""-style include. By default, and in case of a conflict with
+    /// AngledHeaders (i.e. a header matches a regex in both QuotedHeaders and
+    /// AngledHeaders), system headers use <> and non-system headers use "".
+    /// These can match any suffix of the header file in question.
```
[the remark this answers is not in the document]
Comment (kleinesfilmroellchen) [reply clang-tools-extra/clangd/ConfigFragment.h:304]: We're matching against the spelling; at least in SerenityOS this is a necessary distinction: Naming a same-directory header via `"Widget.h"` will result in quotes, but naming the same header via an "absolute" path like `<LibGUI/Widget.h>` should get you an angled include.

**Comment by kleinesfilmroellchen** [comment]
> It's sad that there's not a single codepath that would handle all these, but maybe the perfect is the enemy of the good here.

The code duplication was my reason for looking at the common path. If you're fine with duplicating the code I will go ahead and do that.

**Comment by kleinesfilmroellchen** [comment]
> It's sad that there's not a single codepath that would handle all these, but maybe the perfect is the enemy of the good here.

The code duplication was my reason for looking at the common path. If you're fine with duplicating the code I will go ahead and do that.

I have updated the commit with the review comments and added end-to-end tests.

**Comment by HighCommander4** [comment]
Sorry, I'm still struggling with Github reviews compared to Phabricator... how do I get to an interdiff showing the latest version of the patch compared to the version I reviewed?

**Comment by kleinesfilmroellchen** [comment]
> Sorry, I'm still struggling with Github reviews compared to Phabricator... how do I get to an interdiff showing the latest version of the patch compared to the version I reviewed?

I'm fairly sure that GitHub has no built-in feature for this; I've never used it for my own reviews but it's unfortunate that that is missing of course.

**Comment by ADKaster** [comment]
Yeah seems llvm prefers not to force push PRs, but to keep stacking commits. The squashed commit will have the PR description as the final commit description. The compare button is a bit useless when it has a bunch of changes from main on it (https://github.com/llvm/llvm-project/compare/8a2b22e69f82e3f4caa271b57b998d1c03b21d39..67971ca27ef5e2767aba5cfe2cec1021c4de5ec1)

**Comment by HighCommander4** [comment]
> The compare button is a bit useless when it has a bunch of changes from main on it (https://github.com/llvm/llvm-project/compare/8a2b22e69f82e3f4caa271b57b998d1c03b21d39..67971ca27ef5e2767aba5cfe2cec1021c4de5ec1)

Thanks. Yeah, it looks like the "Compare" link next to the force-push notification in the Conversation view is the sort of thing I'm looking for, except that in this case, the patch was rebased since the previous version, and the rebase and the changes to address the review comments were pushed together, and, unlike Phabricator, Github doesn't seem to know to filter out the former.

**Comment by HighCommander4** [comment]
Anyways, I rebased the old version of the patch locally so that I could generate an interdiff, which I'll attach for reference:

<details>

```diff
diff --git a/clang-tools-extra/clangd/Config.h b/clang-tools-extra/clangd/Config.h
index 12ed1420c93e..3559591fce00 100644
--- a/clang-tools-extra/clangd/Config.h
+++ b/clang-tools-extra/clangd/Config.h
@@ -124,7 +124,7 @@ struct Config {
     // ::). All nested namespaces are affected as well.
     std::vector<std::string> FullyQualifiedNamespaces;
 
-    // List of regexes for inserting certain headers with <> or "".
+    // List of matcher functions for inserting certain headers with <> or "".
     std::vector<std::function<bool(llvm::StringRef)>> QuotedHeaders;
     std::vector<std::function<bool(llvm::StringRef)>> AngledHeaders;
   } Style;
diff --git a/clang-tools-extra/clangd/ConfigFragment.h b/clang-tools-extra/clangd/ConfigFragment.h
index 81474bbd86a5..f0d1080f6c83 100644
--- a/clang-tools-extra/clangd/ConfigFragment.h
+++ b/clang-tools-extra/clangd/ConfigFragment.h
@@ -302,12 +302,16 @@ struct Fragment {
     /// AngledHeaders (i.e. a header matches a regex in both QuotedHeaders and
     /// AngledHeaders), system headers use <> and non-system headers use "".
     /// These can match any suffix of the header file in question.
+    /// Matching is performed against the header text, not its absolute path
+    /// within the project.
     std::vector<Located<std::string>> QuotedHeaders;
     /// List of regexes for headers that should always be included with a
     /// <>-style include. By default, and in case of a conflict with
     /// AngledHeaders (i.e. a header matches a regex in both QuotedHeaders and
     /// AngledHeaders), system headers use <> and non-system headers use "".
     /// These can match any suffix of the header file in question.
+    /// Matching is performed against the header text, not its absolute path
+    /// within the project.
     std::vector<Located<std::string>> AngledHeaders;
   };
   StyleBlock Style;
diff --git a/clang-tools-extra/clangd/Headers.h b/clang-tools-extra/clangd/Headers.h
index e1f3eaa3397b..b91179da253e 100644
--- a/clang-tools-extra/clangd/Headers.h
+++ b/clang-tools-extra/clangd/Headers.h
@@ -9,7 +9,6 @@
 #ifndef LLVM_CLANG_TOOLS_EXTRA_CLANGD_HEADERS_H
 #define LLVM_CLANG_TOOLS_EXTRA_CLANGD_HEADERS_H
 
-#include "Config.h"
 #include "Protocol.h"
 #include "SourceCode.h"
 #include "index/Symbol.h"
diff --git a/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp b/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp
index 6d387fec9b38..bf264a5a44ee 100644
--- a/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp
+++ b/clang-tools-extra/clangd/unittests/CodeCompleteTests.cpp
@@ -11,6 +11,7 @@
 #include "ClangdServer.h"
 #include "CodeComplete.h"
 #include "Compiler.h"
+#include "Config.h"
 #include "Feature.h"
 #include "Matchers.h"
 #include "Protocol.h"
@@ -916,6 +917,41 @@ TEST(CompletionTest, NoIncludeInsertionWhenDeclFoundInFile) {
                           AllOf(named("Y"), Not(insertInclude()))));
 }
 
+TEST(CompletionTest, IncludeInsertionRespectsQuotedAngledConfig) {
+  TestTU TU;
+  TU.ExtraArgs.push_back("-I" + testPath("sub"));
+  TU.AdditionalFiles["sub/bar.h"] = "";
+  auto BarURI = URI::create(testPath("sub/bar.h")).toString();
+
+  Symbol Sym = cls("ns::X");
+  Sym.CanonicalDeclaration.FileURI = BarURI.c_str();
+  Sym.IncludeHeaders.emplace_back(BarURI, 1, Symbol::Include);
+  Annotations Test("int main() { ns::^ }");
+  TU.Code = Test.code().str();
+  auto Results = completions(TU, Test.point(), {Sym});
+  // Default for a local path is quoted include
+  EXPECT_THAT(Results.Completions,
+              ElementsAre(AllOf(named("X"), insertInclude("\"bar.h\""))));
+  {
+    Config C;
+    C.Style.AngledHeaders.push_back(
+        [](auto header) { return header == "bar.h"; });
+    WithContextValue WithCfg(Config::Key, std::move(C));
+    Results = completions(TU, Test.point(), {Sym});
+    EXPECT_THAT(Results.Completions,
+                ElementsAre(AllOf(named("X"), insertInclude("<bar.h>"))));
+  }
+  {
+    Config C;
+    C.Style.QuotedHeaders.push_back(
+        [](auto header) { return header == "bar.h"; });
+    WithContextValue WithCfg(Config::Key, std::move(C));
+    Results = completions(TU, Test.point(), {Sym});
+    EXPECT_THAT(Results.Completions,
+                ElementsAre(AllOf(named("X"), insertInclude("\"bar.h\""))));
+  }
+}
+
 TEST(CompletionTest, IndexSuppressesPreambleCompletions) {
   Annotations Test(R"cpp(
       #include "bar.h"
@@ -1134,8 +1170,8 @@ TEST(CodeCompleteTest, NoColonColonAtTheEnd) {
 }
 
 TEST(CompletionTests, EmptySnippetDoesNotCrash) {
-    // See https://github.com/clangd/clangd/issues/1216
-    auto Results = completions(R"cpp(
+  // See https://github.com/clangd/clangd/issues/1216
+  auto Results = completions(R"cpp(
         int main() {
           auto w = [&](auto &&f) { return f(f); };
           auto f = w([&](auto &&f) {
@@ -1151,18 +1187,18 @@ TEST(CompletionTests, EmptySnippetDoesNotCrash) {
 }
 
 TEST(CompletionTest, Issue1427Crash) {
-    // Need to provide main file signals to ensure that the branch in
-    // SymbolRelevanceSignals::computeASTSignals() that tries to
-    // compute a symbol ID is taken.
-    ASTSignals MainFileSignals;
-    CodeCompleteOptions Opts;
-    Opts.MainFileSignals = &MainFileSignals;
-    completions(R"cpp(
+  // Need to provide main file signals to ensure that the branch in
+  // SymbolRelevanceSignals::computeASTSignals() that tries to
+  // compute a symbol ID is taken.
+  ASTSignals MainFileSignals;
+  CodeCompleteOptions Opts;
+  Opts.MainFileSignals = &MainFileSignals;
+  completions(R"cpp(
       auto f = []() {
         1.0_^
       };
     )cpp",
-                {}, Opts);
+              {}, Opts);
 }
 
 TEST(CompletionTest, BacktrackCrashes) {
diff --git a/clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp b/clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp
index 0eea7ff01b48..efee23003067 100644
--- a/clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp
+++ b/clang-tools-extra/clangd/unittests/ConfigCompileTests.cpp
@@ -555,7 +555,9 @@ TEST_F(ConfigCompileTests, Style) {
       return false;
     };
     EXPECT_TRUE(HeaderFilter("foo.h"));
+    EXPECT_TRUE(HeaderFilter("prefix/foo.h"));
     EXPECT_FALSE(HeaderFilter("bar.h"));
+    EXPECT_FALSE(HeaderFilter("foo.h/bar.h"));
   }
 
   {
```

</details>

**Comment by HighCommander4** [comment]
(In the future, it would greatly help me as a reviewer if rebases are pushed separately from updates to address review comments, so that the latter get their own Compare link.)

**Comment by HighCommander4** [comment]
> > It's sad that there's not a single codepath that would handle all these, but maybe the perfect is the enemy of the good here.
> 
> The code duplication was my reason for looking at the common path. If you're fine with duplicating the code I will go ahead and do that.

+1, please go ahead and wire up the include cleaner support based on Sam's outline; it seems fairly straightforward, the amount of duplication is not much, and we might as well have the config work consistently across all include-insertion features.

The new test + other changes look good as well!

**Comment by travisdowns** [comment]
@kleinesfilmroellchen or @ADKaster it's been quiet for a while, do you think there's any hope for this one still, or is this PR abandoned? Great work to date, BTW, this would be a big win for a project I work on as well (which unconditionally uses <> for internal headers).

**Comment by kleinesfilmroellchen** [comment]
I don't have any energy and the massive reachitecting requested would take me tons of time. Please adopt this. Serenity folks are also waiting fwiw

**Comment by travisdowns** [comment]
> I don't have any energy and the massive reachitecting requested would take me tons of time. Please adopt this.

Thanks, I appreciate the update. Unfortunately I cannot adopt this work at this time.

**Comment by HighCommander4** [comment]
> I don't have any energy and the massive reachitecting requested would take me tons of time.

I certainly didn't intend to request a massive rearchitecting. Sam's comment gave me the impression that hooking up the preferences to include-cleaner would be fairly straightforward.

Anyways, Sam also said:

> Landing this without include-cleaner support is OK with me

so I'm fine with leaving that to a future enhancement.

@kleinesfilmroellchen could you rebase the patch to a recent `main` please (or merge `main` into it)? As far as I can tell, all other comments have been addressed, so if the patch has a green buildkite run on top of recent `main`, I'm happy to merge it without further changes.

**Comment by HighCommander4** [comment]
> @kleinesfilmroellchen could you rebase the patch to a recent `main` please (or merge `main` into it)?

(I just realized there is an "Update branch" button which allows me to do this from the Github UI. Apologies, one year after the project's switch away from Phabricator and I still struggle at times to get the hang of Github...)

**Comment by HighCommander4** [comment]
The rebased patch has green buildkite runs, so I will go ahead and merge it.

**Comment by HighCommander4** [comment]
Thank you @kleinesfilmroellchen for your contribution!

**Comment by ChillerDragon** [comment]
Did this already land in 19.1.7? I am having trouble getting it to work.

My ``~/.clangd`` looks like this just to be sure I also put it in ``~/Desktop/git/ddnet/.clangd`` the repo root.

```yaml
CompileFlags:
  Add: -I.
Diagnostics:
  MissingIncludes: Strict
Style:
  AngledHeaders: ["base/.*", "engine/.*", "game/.*"]
```

:LspInfo in neovim confirms I am running clang 19.1.7

```
LSP configs active in this buffer (bufnr: 2) ~
- Language client log: ~/.local/state/nvim/lsp.log
- Detected filetype: `cpp`
- 1 client(s) attached to this buffer
- Client: `clangd` (id: 2, bufnr: [2])
  root directory:    ~/Desktop/git/ddnet/
  filetypes:         c, cpp, objc, objcpp, cuda, proto
  cmd:               ~/.local/share/nvim/mason/bin/clangd
  version:           `clangd version 19.1.7`
  executable:        true
  autostart:         true
```

When I start to call functions from system.h it automatically included it in quotes ._.
I was expecting all base/.* to be angled.

```C++
#include "base/system.h"
```

**Comment by HighCommander4** [comment]
> Did this already land in 19.1.7?

No, it will appear in clangd 20.

**Pull request merged**